kythe/go/serving/pipeline/nodes/nodes.go

```diff
@@ -207,3 +208,101 @@ func embedSourceKey(src *spb.VName, n *ppb.Node) *ppb.Node {
 		Edge:    n.Edge,
 	}
 }
+
+// Filter is a beam DoFn that emits *ppb.Nodes matching a set of kinds/subkinds.
+// Optionally, each processed node's facts/edges will also be filtered to
+// desired set.
+type Filter struct {
```

Comment:
I'm not sure whether there's some aspect of the existing API that requires us to have a single `ProcessElement` method to do all of these kinds of filtering, but I wonder whether the filtering code would be easier to read and understand if we broke this up into different methods: 
* one for filtering nodes by kind/subkind
* one for filtering node facts
* one for filtering (incident) edges

As it is, there's a lot of shortcut/special cases that make the code harder to read and understand.

Thoughts?

Reply:
There's no Beam requirement for this structure, but I find that it greatly simplifies the client code to have a single `Filter` over nodes.

Reply:
Thanks, this is very helpful and does an excellent job of both describing the semantics and giving useful examples.